src/tfr_control/src/bin_control_server.cpp

```diff
@@ -0,0 +1,82 @@
+#include "bin_control_server.h"
+#include <boost/bind.hpp>
+#include "std_msgs/Float64.h"
+
+
+// I could also pass in the name of the server, but it would end up being a literal
+// string anyway unless it also gets passed into the node, which seems a bit excessive.
+// If it's going to be a literal string we might as well put it as close to its destination
+// as possible so it's clear what it's being used for.
+BinControlServer::BinControlServer(ros::NodeHandle& n, float target_bin_angle) : node{n}, 
+    server{n, "bin_control", boost::bind(&BinControlServer::ControlBin, this, _1), false},
+    TARGET_BIN_ANGLE{target_bin_angle}, signal_mutex{}, bin_task_completed{false}
+{
+    bin_command_publisher = node.advertise<std_msgs::Float64>(
+            "bin_position_controller/command", 100);
+    server.start();
+}
+
+// Action Server callback function
+void BinControlServer::ControlBin(const tfr_msgs::BinGoalConstPtr& goal)
+{
+    std_msgs::Float64 command;
+    tfr_msgs::BinResult result;
+
+    if (goal->command_code == goal->RAISE_BIN)
+    {
+        command.data = TARGET_BIN_ANGLE;
+        bin_command_publisher.publish(command);
+    }
+    else if (goal->command_code == goal->LOWER_BIN)
+    {
+        command.data = 0;
+        bin_command_publisher.publish(command);
+    }
+    else
+    {
+        ROS_WARN("Bin Controller received an invalid control code: %d", goal->command_code);
+        result.return_code = result.ERROR_ENCOUNTERED;
+        server.setAborted(result);
+        return;
+    }
+
+    // unlocked by the wait() call
+    std::unique_lock<std::mutex> signal_lock(signal_mutex);
+    while (!bin_task_completed)
+    {
+        // TODO: Instead of using the signal.wait() semantics, we could check for preemptions in a 
+        //       polling loop here, but I don't think there's an easy way to tell the hardware_interface
+        //       to simply stop moving the bin; it takes in a target angle and simply goes until it gets 
+        //       there. If we think supporting preemption is important for this component, we can 
+        //       investigate further options.
+
+        // http://en.cppreference.com/w/cpp/thread/condition_variable 
+        // https://stackoverflow.com/questions/16350473/why-do-i-need-stdcondition-variable
+        // wait() will release the lock. Sometimes the signal will be notified by accident, so
+        // a check on bin_task_completed is required. wait() reaquires the lock after it unblocks.
+        signal.wait(signal_lock);
```

Comment:
From what I understand from you code, 
1. client sets goal
2.  server receives it and publishes on topic
3.  block until completed.
  - controller.cpp sets condition variable when it finishes. 
4.  return result.

Thats fine, but it took me too long to get there from reading your code. 

You need to take a few steps to make this more readable and maintainable for future readers:
- Extract all the blocking synchronization code into a separate method: e.g. `wait_for_bin()`
  - In that method tell the readers what's happening, who we are blocking on, and what events trigger this.
  - I think at least 10 minutes of reading could be shaved of "reader comprehension time" if we did this, synchronization usually needs a little extra care in documentation if you want to maintain it..
- If we aren't supporting preemption, explain why in the class header, and in the description of the method.

Reply:
Sure.  Yeah, you've got it right.